contracts/toCKB-typescript/src/utils/tools.rs

```diff
@@ -191,6 +195,116 @@ pub fn verify_btc_witness(
     }
 }
 
+pub fn verify_header_is_on_main_chain(
+    header: &BlockHeader,
+    cell_dep_index_list: &[u8],
+) -> Result<(), Error> {
+    let dep_data = load_cell_data(cell_dep_index_list[0].into(), Source::CellDep)?;
+    debug!("dep data is {:?}", &dep_data);
+    if EthCellDataReader::verify(&dep_data, false).is_err() {
+        return Err(Error::InvalidWitness);
+    }
+    let eth_cell_data_reader = EthCellDataReader::new_unchecked(&dep_data);
+    debug!("eth_cell_data_reader: {:?}", eth_cell_data_reader);
+    let tail_raw = eth_cell_data_reader
+        .headers()
+        .main()
+        .get_unchecked(eth_cell_data_reader.headers().main().len() - 1)
+        .raw_data();
+    if HeaderInfoReader::verify(&tail_raw, false).is_err() {
+        return Err(Error::EthHeadersDataInvalid);
+    }
+    let tail_info_reader = HeaderInfoReader::new_unchecked(tail_raw);
+    let tail_info_raw = tail_info_reader.header().raw_data();
+    let tail: BlockHeader = rlp::decode(tail_info_raw.to_vec().as_slice()).unwrap();
+    if header.number > tail.number {
+        return Err(Error::HeaderIsNotOnMainChain);
+    }
+    let offset = (tail.number - header.number) as usize;
+    if offset > eth_cell_data_reader.headers().main().len() - 1 {
+        return Err(Error::HeaderIsNotOnMainChain);
+    }
+    let target_raw = eth_cell_data_reader
+        .headers()
+        .main()
+        .get_unchecked(eth_cell_data_reader.headers().main().len() - 1 - offset)
+        .raw_data();
+    let target_info_reader = HeaderInfoReader::new_unchecked(target_raw);
+    debug!(
+        "main chain hash: {:?}, witness header hash: {:?}",
+        hex::encode(target_info_reader.hash().raw_data()),
+        hex::encode(header.hash.expect("invalid hash").0.as_bytes())
+    );
+    if target_info_reader.hash().raw_data() != header.hash.expect("invalid hash").0.as_bytes() {
+        return Err(Error::HeaderIsNotOnMainChain);
+    }
+    Ok(())
+}
+
+pub fn verify_eth_witness(
+    data: &ToCKBCellDataView,
+    proof: &[u8],
+    cell_dep_index_list: &[u8],
+) -> Result<EthExtraView, Error> {
```

Comment:
two questions here:
1. Miss lock amount verification?
2. Dose eth-transfer tx have a receipt?
If dosen't, we should verify the tx-structure rather than receipt